### ISSUE TITLE
[QUESTION] Is FileStorage in encrypted datafile also encrypted?

### ISSUE THREAD

**Issue by Dieveral** [litedb-org/LiteDB#1997]
I have files stored in FileStorage. FileStorage could be huge. Datafile with this FileStorage is encrypted with password. Does it mean FileStorage is also encrypted?

**Comment by JKamsker** [comment]
Yes. With a password, LiteDB encrypts the datafile (and its `-log.db`) page by page with AES. FileStorage isn't stored anywhere special: it's just two ordinary collections, `_files` and `_chunks`, in those same pages. So file content is encrypted exactly like your documents.

A quick check on 5.0.21: I uploaded a ~440 KB file made of a repeated ASCII marker. Without a password, the raw `.db` contains the marker about 20,000 times. With `Password=secret` it contains it zero times, and `Download` still returns the original bytes.

Keep in mind that content is decrypted when you read it through LiteDB (`OpenRead`/`Download`), so any copies you write to disk yourself aren't protected. Also, if you ever call `Rebuild()` on that database, pass the password via `RebuildOptions` (#2449). Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.

**Update (#2525):** Since v5.0.19, `Rebuild()` fails with `Detected loop in FindAll(...)` on collections that have a secondary index and more than about 2,500 documents (#2525). Your data stays intact, but the file isn't rebuilt and a `-temp.db` file is left behind.

Until that's fixed, drop the secondary indexes before rebuilding and recreate them afterwards:

```csharp
// for each secondary index on the collection:
col.DropIndex("Name");
db.Rebuild();                     // or db.Rebuild(new RebuildOptions { Password = ... })
col.EnsureIndex(x => x.Name);
```

We tested this on 5.0.16, 5.0.21 and the current 6.x prerelease; all data and indexes are kept.
